Implement plugins infrastructure in User_LDAP

viniciuscb opened the pull request:
This PR is related to issue https://github.com/nextcloud/server/issues/3476, following the suggestion from @blizzz : It implements a plugin infra-structure that allows other apps to implement other actions in LDAP, like setDisplayName and createUser. 

It also allows other apps to change the NC backends order, so that the createUser might be handled by the LDAP Backend, instead of the default Database backend. 

We are developing a working app which uses the infrastructure. It's still not prepared for production, but fully works for createUser, setDisplayName, changeAvatar and deleteUser.

https://gitlab.com/eita/rios/user_ldap_extended/tags/alpha1

LukasReschke commented on apps/user_ldap/lib/ILDAPPlugin.php:
```diff
@@ -0,0 +1,104 @@
+	 * @return int|bool
+	 */
+	public function countUsers();
+}
```
Missing newline at EoF :)

LukasReschke commented on apps/user_ldap/lib/LDAPProvider.php:
```diff
@@ -184,4 +188,31 @@ public function flagRecord($uid) {
+	 */
+	public function getLDAPDisplayNameField($uid) {
+		if(!$this->backend->userExists($uid)){
+			throw new \Exception('User id not found in LDAP');
```
Can I have a test for the error condition as well? See https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.expectedException

viniciuscb replied:
We will rather implement @blizzz suggestion https://github.com/nextcloud/server/pull/5321#discussion_r121937322

LukasReschke commented on apps/user_ldap/lib/LDAPProvider.php:
```diff
@@ -184,4 +188,31 @@ public function flagRecord($uid) {
+	 */
+	public function getLDAPEmailField($uid) {
+		if(!$this->backend->userExists($uid)){
+			throw new \Exception('User id not found in LDAP');
```
Can I have a test for the error condition as well? See https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.expectedException

viniciuscb replied:
We will rather implement @blizzz suggestion https://github.com/nextcloud/server/pull/5321#discussion_r121937322

LukasReschke commented on apps/user_ldap/lib/PluginManager.php:
```diff
@@ -0,0 +1,182 @@
+		}
+	}
+
+	public static function implementsActions($actions){
```
Space before {

viniciuscb replied:
Ok.

LukasReschke commented on apps/user_ldap/lib/PluginManager.php:
```diff
@@ -0,0 +1,182 @@
+	}
+
+
+}
```
Missing newline at EoF

viniciuscb replied:
Ok.

LukasReschke commented on apps/user_ldap/lib/User_LDAP.php:
```diff
@@ -71,6 +73,10 @@ public function __construct(Access $access, IConfig $ocConfig, INotificationMana
 	 * @return boolean either the user can or cannot
 	 */
 	public function canChangeAvatar($uid) {
+		if (PluginManager::implementsActions(Backend::PROVIDE_AVATAR)) {
```
Can we use Dependency Injection instead so we can easier mock this for tests?

LukasReschke replied:
Also valid for the other occurrences below of course.

viniciuscb replied:
Ok. We are now trying to implement it. As we said below, we're having difficulties to know how to inject the new class pluginManager and use it in user_LDAP as a single instance.

We've registered the class as a new service in Application.php, but we don't know how to retrieve it or how to retrieve the app container to do the query inside user_LDAP.

Any hints on how to do that? Thanks!

LukasReschke commented on apps/user_ldap/lib/User_LDAP.php:
```diff
@@ -549,4 +593,17 @@ public function getNewLDAPConnection($uid) {
+		}
+		return false;
+	}
+}
```
Missing newline at Eof

LukasReschke commented on apps/user_ldap/tests/LDAPPluginDummy.php:
```diff
@@ -0,0 +1,51 @@
+ * User: vinicius
+ * Date: 09/06/17
+ * Time: 12:10
+ */
```
License header needs a fix :)

LukasReschke replied:
Kudos for the tests though! 🚀

LukasReschke commented on apps/user_ldap/tests/LDAPPluginDummy.php:
```diff
@@ -0,0 +1,51 @@
+	public function countUsers() {
+		return null;
+	}
+}
```
Missing newline

LukasReschke commented on apps/user_ldap/tests/LDAPPluginTest.php:
```diff
@@ -0,0 +1,216 @@
+		PluginManager::reset();
+		return $res;
+	}
+}
```
Missing newline

viniciuscb replied:
Ok.

LukasReschke reviewed (dismissed):
I'm not deep enough into the LDAP code to judge this properly but on a first look this seems promising. Some nitpicks below :)

blizzz commented on apps/user_ldap/appinfo/app.php:
```diff
@@ -27,6 +29,8 @@
  *
  */
 
+use OCA\User_LDAP\PluginManager;
```
not used in this file

blizzz commented on apps/user_ldap/lib/User_LDAP.php:
```diff
@@ -149,6 +155,10 @@ public function getLDAPUserByLoginName($loginName) {
 	 * @return false|string
 	 */
 	public function checkPassword($uid, $password) {
+		if (PluginManager::implementsActions(Backend::CHECK_PASSWORD)) {
```
not sure we should enable overwriting of the most central methods. In that case, a Plugin is better writing off its backend itself. But perhaps I am missing the  motivation to allow it?

blizzz reviewed (commented):
Initially I thought making the plugin manager available through the OCP namespace, but this is too close to the backends implementation, so 👍 for keeping it in OCA\User_LDAP.

Thus, it's kept relatively simple. I have some remarks and questions, but these are implementation details.  The big picture looks good.

blizzz commented:
fyi @nextcloud/ldap

viniciuscb commented on apps/user_ldap/lib/PluginManager.php:
```diff
@@ -0,0 +1,182 @@
+		'deleteUser' => null
+	);
+
+	public static function reset() {
```
Agreed. 

We are trying to do that, but we're having difficulties to know how to inject the new class _pluginManager_ and use it in _user_LDAP_ as a single instance. 

We've registered the class as a new service in _Application.php_, but we don't know how to retrieve it or how to retrieve the app container to do the query inside _user_LDAP_.

blizzz replied:
This file was not added, yet. 

The LDAP Backend was written before the AppFramework was in place and it is not totally adjusted to it. Currently, the initialization happens within the starting points, e.g. in app.php. Create an instance there and pass it on as constructor parameter to the classes that needs to make use of it.

viniciuscb commented on apps/user_ldap/lib/LDAPProvider.php:
```diff
@@ -184,4 +188,31 @@ public function flagRecord($uid) {
+	 * @return string the display name field
+	 * @throws \Exception if user id was not found in LDAP
+	 */
+	public function getLDAPDisplayNameField($uid) {
```
Ok. We're implementing it through a new class instance to make it more elegant.

viniciuscb replied:
@blizzz We've noticed also that the group implementation into a ldap plugin is also necessary. So I've implemented it and in this case there were big changes. I think maybe we can review the actual alterations and then put this suggestion for LDAPProvider into a new PR.

blizzz commented:
@viniciuscb I guess are still in development, since I saw commits, but no further comments. Just want to inform you, that i am on parental leave for a month - cannot promise reviews in that time.

viniciuscb commented:
@blizzz @LukasReschke We've noticed that we missed the group implementation into the ldap plugin, for the operations of creating and deleting groups, and adding and removing user to group (for a LDAP group). 

So We've implemented it and in this case there are more changes and new files.

I've applied all the suggestions and corrections you've made but one (The one by @blizz as Jun 14. I've suggested him to do this one in a separated PR, as it is an improvement, and not a correction). 

The tests were also implemented and I've conferred them through code coverage. 

Now we can continue the review process.

blizzz commented on lib/public/IGroupManager.php:
```diff
@@ -66,6 +66,13 @@ public function addBackend($backend);
+	 * @return \OCP\GroupInterface[]
+	 * @since 13.0.0
+	 */
+	public function getBackends();
```
how critical is it? This extends the API and would make potential backends incompatible, if not adapted. Playing the conservative card here, but if it's needed, let's get it in.

viniciuscb replied:
As in other places, we wanted to make IGroupManager as similar as possible to the existing IUserManager. Since there is a way to return active backends for users in its API, we thought it would be way more elegant to offer the same method to groups.

We need it in LDAPProvider to have the backend for a given group.

blizzz reviewed (changes requested):
I am back… Impressive work! For now i did a formal check, and a couple of questions came up, and some minor things.

blizzz commented on apps/user_ldap/lib/LDAPProvider.php:
```diff
@@ -184,4 +188,31 @@ public function flagRecord($uid) {
+	 * @return string the display name field
+	 * @throws \Exception if user id was not found in LDAP
+	 */
+	public function getLDAPDisplayNameField($uid) {
```
Yup… meanwhile I do not see this really critical or anything. Fine with setters, perhaps the optional display name field should get one, too, or not. I am OK with the state.

blizzz commented:
@viniciuscb heyo, would like to ask what the current state of the PR is?

viniciuscb commented:
Hi @blizzz I'll take a look and answer asap.

viniciuscb commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -1067,17 +1076,113 @@ public function groupExists($gid) {
 	 * @return Access instance of Access for LDAP interaction
 	 */
-	public function getLDAPAccess() {
+	public function getLDAPAccess($gid) {
```
That is implemented the same way it is in apps/user_ldap/lib/User_LDAP.php . I had to include this parameter after I had to create the interface IGroupLDAP. I had to create this iGroupLDAP interface because I needed to implement the method LDAPProvider.getGroupLDAPConnection($gid) to get the specific LDAP connection for a given LDAP group. 

This implementation was almost a copy of the implementation that was already being done for the user. Only that for the user we had IUserLDAP and for the group we did not had IGroupLDAP, I had to implement it to be used in LDAPProvider.

viniciuscb commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -1067,17 +1076,113 @@ public function groupExists($gid) {
+			if ($ret = $this->groupPluginManager->createGroup($gid)) {
+				$this->access->connection->clearCache();
+				#update the list of groups in nextcloud internal db
+				$groups = $this->getGroups();
```
I've done it an alternative way. Going in next commit...

blizzz commented:
@viniciuscb we'll have feature freeze for 13 on Friday. Do you need any help?

viniciuscb commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -1067,17 +1076,113 @@ public function groupExists($gid) {
+			}
+			return $ret;
+		}
+		return null;
```
ok!

viniciuscb commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -1067,17 +1076,113 @@ public function groupExists($gid) {
+			}
+			return $ret;
+		}
+		return null;
```
should not throw Exception like the previous one?

blizzz replied:
I don't see to which function it refers anymore, Github does not show more context 😒 . Basically it depends if success is returned (→ bool). In this case, both success or failure are expected. What the prior check is about might also have a word saying in, in the referred commit the condition was like it just always be fulfilled (based on my comment).  The managing bit should be less intrusive. If the backend plugin decides to be the one any only that can add groups for instance, it would be OK to throw an exception to prevent the nextcloud core manager to ask the local backend. To give one example.

viniciuscb commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -1067,17 +1076,113 @@ public function groupExists($gid) {
+			}
+			return $ret;
+		}
+		return null;
```
should not throw Exception like the previous one?

viniciuscb commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -1067,17 +1076,113 @@ public function groupExists($gid) {
+			}
+			return $ret;
+		}
+		return null;
```
should not throw Exception like the previous one?

viniciuscb commented on apps/user_ldap/lib/Group_Proxy.php:
```diff
@@ -203,6 +258,17 @@ public function implementsActions($actions) {
 	 */
 	public function getLDAPAccess($gid) {
-		return $this->handleRequest($gid, 'getLDAPAccess', []);
+		return $this->handleRequest($gid, 'getLDAPAccess', [$gid]);
```
explained above.

viniciuscb commented on apps/user_ldap/lib/ILDAPGroupPlugin.php:
```diff
@@ -0,0 +1,89 @@
+
+	/**
+	 * Check if plugin implements actions
+	 * @param int $actions bitwise-or'ed actions
```
ok!

viniciuscb commented on apps/user_ldap/lib/ILDAPUserPlugin.php:
```diff
@@ -0,0 +1,95 @@
+
+	/**
+	 * Check if plugin implements actions
+	 * @param int $actions bitwise-or'ed actions
```
ok!

viniciuscb commented on apps/user_ldap/lib/ILDAPUserPlugin.php:
```diff
@@ -0,0 +1,95 @@
+	/**
+	 * Create a new user in LDAP Backend
+	 *
+	 * @param string $uid The username of the user to create
```
ok!

viniciuscb commented on apps/user_ldap/lib/LDAPProvider.php:
```diff
@@ -115,10 +151,24 @@ public function sanitizeDN($dn) {
+	/**
+	 * Return a new LDAP connection resource for the specified user.
+	 * The connection must be closed manually.
+	 * @param string $uid user id
```
ok!

viniciuscb commented on apps/user_ldap/lib/LDAPProvider.php:
```diff
@@ -115,10 +151,24 @@ public function sanitizeDN($dn) {
+	 * The connection must be closed manually.
+	 * @param string $uid user id
+	 * @return resource of the LDAP connection
+	 * @throws \Exception if user id was not found in LDAP
```
ok!

viniciuscb commented:
@blizzz please take a look. If there is some additional change or adjustment to do, please tell me, I will put this over my priorities so we can have this feature in PR 13.

We implemented a plugin that I am not sure would work in any installation, because User_LDAP can be configurated in much ways (as LDAP do). Maybe it assumes a specific configuration, for instance, the association of groups to members can have several types, not all are implemented. But if you are interested, you can try it, we can pass you the coordinates.

blizzz commented:
Intermediate feedback (more will come later…): 

## Failing tests:

* 9 tests failing in OCA\User_LDAP\Tests\UserLDAPPluginTest (https://drone.nextcloud.com/nextcloud/server/2526/91 )

## Formality:

Some commits were not signed off!
Missing signatures on:
- 208ce1604c06996f445c4e741aa8c973d991fa86
- fa80172df28ef9bf24cdfe4f4259c59bff5e3f3b
- a8e732ad78ca36a757951d8220d75fce0e89d71d
- da918e87138bc9016e8b973c6fac1ffac2fca4f2
- b212912169eb40585c455b0e16357bafc365635b
- d5aee379818ddf441d1ba76e4a07598fa0c581dd
- bc3f9a16fec17d63c79407949816626425f1f319

(https://drone.nextcloud.com/nextcloud/server/2526/77 )

blizzz commented on apps/user_ldap/lib/UserPluginManager.php:
```diff
@@ -0,0 +1,208 @@
+
+	/**
+	 * Registers a group plugin that may implement some actions, overriding User_LDAP's user actions.
+	 * @param ILDAPGroupPlugin $plugin
```
ILDAPGroupPlugin → …User…, see patch

blizzz reviewed (commented):
> @blizzz please take a look. If there is some additional change or adjustment to do, please tell me, I will put this over my priorities so we can have this feature in PR 13.

Smaller changes, for all of them please see the patches below. I'd just ask you to update the autoloader (command below), next to confirming and applying the patches and getting the sign-offs ^, then we're good and you get my :+1:.

> We implemented a plugin that I am not sure would work in any installation, because User_LDAP can be configurated in much ways (as LDAP do). Maybe it assumes a specific configuration, for instance, the association of groups to members can have several types, not all are implemented. But if you are interested, you can try it, we can pass you the coordinates.

Is it possible to have this as an public example? It could go to the documentation, in a blog, or a repo of it's own.

> My IDE says to me that I should also add more methods to interfaces, but here I tested this code, it is working properly. The methods are:

> \OCP\GroupInterface:createGroup

> \OCP\GroupInterface:getGroupDetails

It's correct not adding them. Both methods are optional and should not go into the interface.

* run `bash build/autoloaderchecker.sh` and commit results to have the autoloader updated
* Integration tests are not running through. Need to register servies in the abstract class, and also UserCleanUp Tests needs an adjustement. use following patch. (Ideally the app.php content would go to the Application class, that's not in scope however, i can deal with it later)

```diff
diff --git a/apps/user_ldap/tests/Integration/AbstractIntegrationTest.php b/apps/user_ldap/tests/Integration/AbstractIntegrationTest.php
index 84e1e6b458..8f9104999d 100644
--- a/apps/user_ldap/tests/Integration/AbstractIntegrationTest.php
+++ b/apps/user_ldap/tests/Integration/AbstractIntegrationTest.php
@@ -68,6 +68,13 @@ abstract class AbstractIntegrationTest {
     * the LDAP backend.
     */
    public function init() {
+       \OC::$server->registerService('LDAPUserPluginManager', function() {
+           return new \OCA\User_LDAP\UserPluginManager();
+       });
+       \OC::$server->registerService('LDAPGroupPluginManager', function() {
+           return new \OCA\User_LDAP\GroupPluginManager();
+       });
+
        $this->initLDAPWrapper();
        $this->initConnection();
        $this->initUserManager();
diff --git a/apps/user_ldap/tests/Integration/Lib/User/IntegrationTestUserCleanUp.php b/apps/user_ldap/tests/Integration/Lib/User/IntegrationTestUserCleanUp.php
index ce81b9a26d..10f0968efb 100644
--- a/apps/user_ldap/tests/Integration/Lib/User/IntegrationTestUserCleanUp.php
+++ b/apps/user_ldap/tests/Integration/Lib/User/IntegrationTestUserCleanUp.php
@@ -46,7 +46,7 @@ class IntegrationTestUserCleanUp extends AbstractIntegrationTest {
        $this->mapping->clear();
        $this->access->setUserMapper($this->mapping);
 
-       $userBackend  = new User_LDAP($this->access, \OC::$server->getConfig(), \OC::$server->getNotificationManager(), \OC::$server->getUserSession());
+       $userBackend  = new User_LDAP($this->access, \OC::$server->getConfig(), \OC::$server->getNotificationManager(), \OC::$server->getUserSession(), \OC::$server->query('LDAPUserPluginManager'));
        \OC_User::useBackend($userBackend);
    }
 
```

mostly smaller fixes as described in the comments:

```diff
diff --git a/apps/user_ldap/lib/GroupPluginManager.php b/apps/user_ldap/lib/GroupPluginManager.php
index 6250ae6be0..d242da8b39 100644
--- a/apps/user_ldap/lib/GroupPluginManager.php
+++ b/apps/user_ldap/lib/GroupPluginManager.php
@@ -63,7 +63,7 @@ class GroupPluginManager {
 
 	/**
 	 * Signal if there is a registered plugin that implements some given actions
-	 * @param int $action Actions defined in \OC\Group\Backend, like Backend::REMOVE_FROM_GROUP
+	 * @param int $actions Actions defined in \OC\Group\Backend, like Backend::REMOVE_FROM_GROUP
 	 * @return bool
 	 */
 	public function implementsActions($actions) {
diff --git a/apps/user_ldap/lib/Group_LDAP.php b/apps/user_ldap/lib/Group_LDAP.php
index 39519cc462..7f226b49b0 100644
--- a/apps/user_ldap/lib/Group_LDAP.php
+++ b/apps/user_ldap/lib/Group_LDAP.php
@@ -39,9 +39,9 @@
 namespace OCA\User_LDAP;
 
 use OC\Cache\CappedMemoryCache;
-use OC\Group\Backend;
+use OCP\GroupInterface;
 
-class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLDAP {
+class Group_LDAP extends BackendUtility implements GroupInterface, IGroupLDAP {
 	protected $enabled = false;
 
 	/**
@@ -865,7 +865,7 @@ class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLD
 	 * @return int|bool
 	 */
 	public function countUsersInGroup($gid, $search = '') {
-		if ($this->groupPluginManager->implementsActions(Backend::COUNT_USERS)) {
+		if ($this->groupPluginManager->implementsActions(GroupInterface::COUNT_USERS)) {
 			return $this->groupPluginManager->countUsersInGroup($gid, $search);
 		}
 
@@ -1079,7 +1079,7 @@ class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLD
 	* compared with \OC\Group\Backend::CREATE_GROUP etc.
 	*/
 	public function implementsActions($actions) {
-		return (bool)((\OC\Group\Backend::COUNT_USERS |
+		return (bool)((GroupInterface::COUNT_USERS |
 				$this->groupPluginManager->getImplementedActions()) & $actions);
 	}
 
@@ -1098,7 +1098,7 @@ class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLD
 	 * @throws \Exception
 	 */
 	public function createGroup($gid) {
-		if ($this->groupPluginManager->implementsActions(Backend::CREATE_GROUP)) {
+		if ($this->groupPluginManager->implementsActions(GroupInterface::CREATE_GROUP)) {
 			if ($dn = $this->groupPluginManager->createGroup($gid)) {
 				//updates group mapping
 				$this->access->dn2ocname($dn, $gid, false);
@@ -1116,7 +1116,7 @@ class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLD
 	 * @throws \Exception
 	 */
 	public function deleteGroup($gid) {
-		if ($this->groupPluginManager->implementsActions(Backend::DELETE_GROUP)) {
+		if ($this->groupPluginManager->implementsActions(GroupInterface::DELETE_GROUP)) {
 			if ($ret = $this->groupPluginManager->deleteGroup($gid)) {
 				#delete group in nextcloud internal db
 				$this->access->getGroupMapper()->unmap($gid);
@@ -1135,9 +1135,9 @@ class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLD
 	 * @throws \Exception
 	 */
 	public function addToGroup($uid, $gid) {
-		if ($this->groupPluginManager->implementsActions(Backend::ADD_TO_GROUP)) {
+		if ($this->groupPluginManager->implementsActions(GroupInterface::ADD_TO_GROUP)) {
 			if ($ret = $this->groupPluginManager->addToGroup($uid, $gid)) {
-				#$this->access->connection->clearCache();
+				$this->access->connection->clearCache();
 			}
 			return $ret;
 		}
@@ -1152,9 +1152,9 @@ class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLD
 	 * @throws \Exception
 	 */
 	public function removeFromGroup($uid, $gid) {
-		if ($this->groupPluginManager->implementsActions(Backend::REMOVE_FROM_GROUP)) {
+		if ($this->groupPluginManager->implementsActions(GroupInterface::REMOVE_FROM_GROUP)) {
 			if ($ret = $this->groupPluginManager->removeFromGroup($uid, $gid)) {
-				#$this->access->connection->clearCache();
+				$this->access->connection->clearCache();
 			}
 			return $ret;
 		}
@@ -1168,10 +1168,10 @@ class Group_LDAP extends BackendUtility implements \OCP\GroupInterface, IGroupLD
 	 * @throws \Exception
 	 */
 	public function getGroupDetails($gid) {
-		if ($this->groupPluginManager->implementsActions(Backend::GROUP_DETAILS)) {
+		if ($this->groupPluginManager->implementsActions(GroupInterface::GROUP_DETAILS)) {
 			return $this->groupPluginManager->getGroupDetails($gid);
 		}
-		throw new \Exception('Could not get group details in LDAP backend.');
+		throw new \Exception('Could not get	 group details in LDAP backend.');
 	}
 
 	/**
diff --git a/apps/user_ldap/lib/ILDAPGroupPlugin.php b/apps/user_ldap/lib/ILDAPGroupPlugin.php
index 468424a560..3901b00738 100644
--- a/apps/user_ldap/lib/ILDAPGroupPlugin.php
+++ b/apps/user_ldap/lib/ILDAPGroupPlugin.php
@@ -39,14 +39,14 @@ interface ILDAPGroupPlugin {
 	 * @param string $gid
 	 * @return string|null The group DN if group creation was successful.
 	 */
-    public function createGroup($gid);
+	public function createGroup($gid);
 
 	/**
 	 * delete a group
 	 * @param string $gid gid of the group to delete
 	 * @return bool
 	 */
-    public function deleteGroup($gid);
+	public function deleteGroup($gid);
 
 	/**
 	 * Add a user to a group
@@ -56,7 +56,7 @@ interface ILDAPGroupPlugin {
 	 *
 	 * Adds a user to a group.
 	 */
-    public function addToGroup($uid, $gid);
+	public function addToGroup($uid, $gid);
 
 	/**
 	 * Removes a user from a group
@@ -66,7 +66,7 @@ interface ILDAPGroupPlugin {
 	 *
 	 * removes the user from a group.
 	 */
-    public function removeFromGroup($uid, $gid);
+	public function removeFromGroup($uid, $gid);
 
 	/**
 	 * get the number of all users matching the search string in a group
@@ -74,14 +74,14 @@ interface ILDAPGroupPlugin {
 	 * @param string $search
 	 * @return int|false
 	 */
-    public function countUsersInGroup($gid, $search = '');
+	public function countUsersInGroup($gid, $search = '');
 
 	/**
 	 * get an array with group details
 	 * @param string $gid
 	 * @return array|false
 	 */
-    public function getGroupDetails($gid);
+	public function getGroupDetails($gid);
 
 
 
diff --git a/apps/user_ldap/lib/LDAPProvider.php b/apps/user_ldap/lib/LDAPProvider.php
index 060c478ae3..c75d85af47 100644
--- a/apps/user_ldap/lib/LDAPProvider.php
+++ b/apps/user_ldap/lib/LDAPProvider.php
@@ -42,10 +42,13 @@ class LDAPProvider implements ILDAPProvider, IDeletionFlagSupport {
 	private $logger;
 	private $helper;
 	private $deletedUsersIndex;
-	
+
 	/**
 	 * Create new LDAPProvider
+	 *
 	 * @param \OCP\IServerContainer $serverContainer
+	 * @param Helper $helper
+	 * @param DeletedUsersIndex $deletedUsersIndex
 	 * @throws \Exception if user_ldap app was not enabled
 	 */
 	public function __construct(IServerContainer $serverContainer, Helper $helper, DeletedUsersIndex $deletedUsersIndex) {
diff --git a/apps/user_ldap/lib/UserPluginManager.php b/apps/user_ldap/lib/UserPluginManager.php
index 374e545f4f..316abf5083 100644
--- a/apps/user_ldap/lib/UserPluginManager.php
+++ b/apps/user_ldap/lib/UserPluginManager.php
@@ -51,7 +51,8 @@ class UserPluginManager {
 
 	/**
 	 * Registers a group plugin that may implement some actions, overriding User_LDAP's user actions.
-	 * @param ILDAPGroupPlugin $plugin
+	 *
+	 * @param ILDAPUserPlugin $plugin
 	 */
 	public function register(ILDAPUserPlugin $plugin) {
 		$respondToActions = $plugin->respondToActions();
@@ -70,7 +71,7 @@ class UserPluginManager {
 
 	/**
 	 * Signal if there is a registered plugin that implements some given actions
-	 * @param int $action Actions defined in \OC\User\Backend, like Backend::CREATE_USER
+	 * @param int $actions Actions defined in \OC\User\Backend, like Backend::CREATE_USER
 	 * @return bool
 	 */
 	public function implementsActions($actions) {
@@ -80,7 +81,7 @@ class UserPluginManager {
 	/**
 	 * Create a new user in LDAP Backend
 	 *
-	 * @param string $uid The username of the user to create
+	 * @param string $username The username of the user to create
 	 * @param string $password The password of the new user
 	 * @return bool
 	 * @throws \Exception
```diff

viniciuscb committed:
Plugins infrastructure in User_LDAP

Signed-off-by: Vinicius Cubas Brand <viniciuscb@gmail.com>

viniciuscb commented:
Hi @blizzz, I am having trouble in signing previous commits. Do you have some reference about that?

I was thinking in an alternative: because all commits in this PR belongs to me, I thought in getting all these differences (with rsync -Ca) and group all them in just one signed commit. After that, force push into this branch. Do you think that's ok?

viniciuscb commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -1067,17 +1076,114 @@ public function groupExists($gid) {
+	public function addToGroup($uid, $gid) {
+		if ($this->groupPluginManager->implementsActions(Backend::ADD_TO_GROUP)) {
+			if ($ret = $this->groupPluginManager->addToGroup($uid, $gid)) {
+				#$this->access->connection->clearCache();
```
Agreed. The best option is to extend clearCache() (or add another method) to clear just one entry from the cache. In this case (and below), would have to unset two keys, 

```php
'inGroup'.$uid.':'.$gid
'inGroup-members:'.$gid
```

and these caches would be reconstructed in the next call to inGroup(). The same applies for your comment below.

viniciuscb committed:
User_LDAP plugins: smaller fixes

Signed-off-by: Vinicius Cubas Brand <viniciuscb@gmail.com>

blizzz commented:
@viniciuscb tests are failing: 

> Class 'OCA\User_LDAP\Backend' not found in /drone/src/github.com/nextcloud/server/apps/user_ldap/lib/GroupPluginManager.php on line 80

* Backend must be replaced with GroupInterface  (also the usages in the following methods)

Same for GroupLDAPPluginTest.php (https://drone.nextcloud.com/nextcloud/server/2590/81)

> The autoloaders are not up to date
Please run: bash build/autoloaderchecker.sh
And commit the result

^ this :)

viniciuscb commented:
@blizzz When I run the autoloader, it generates a composer.phar file only. Does not change any already included file. I am commiting it into a separate commit.

blizzz commented:
Well, OK, then add, this is what's generated for me and should not differ anyhow :)

```diff
diff --git a/apps/user_ldap/composer/composer/autoload_classmap.php b/apps/user_ldap/composer/composer/autoload_classmap.php
index 6558af0021..0962fe2c4e 100644
--- a/apps/user_ldap/composer/composer/autoload_classmap.php
+++ b/apps/user_ldap/composer/composer/autoload_classmap.php
@@ -24,9 +24,13 @@ return array(
     'OCA\\User_LDAP\\Exceptions\\ConstraintViolationException' => $baseDir . '/../lib/Exceptions/ConstraintViolationException.php',
     'OCA\\User_LDAP\\Exceptions\\NotOnLDAP' => $baseDir . '/../lib/Exceptions/NotOnLDAP.php',
     'OCA\\User_LDAP\\FilesystemHelper' => $baseDir . '/../lib/FilesystemHelper.php',
+    'OCA\\User_LDAP\\GroupPluginManager' => $baseDir . '/../lib/GroupPluginManager.php',
     'OCA\\User_LDAP\\Group_LDAP' => $baseDir . '/../lib/Group_LDAP.php',
     'OCA\\User_LDAP\\Group_Proxy' => $baseDir . '/../lib/Group_Proxy.php',
     'OCA\\User_LDAP\\Helper' => $baseDir . '/../lib/Helper.php',
+    'OCA\\User_LDAP\\IGroupLDAP' => $baseDir . '/../lib/IGroupLDAP.php',
+    'OCA\\User_LDAP\\ILDAPGroupPlugin' => $baseDir . '/../lib/ILDAPGroupPlugin.php',
+    'OCA\\User_LDAP\\ILDAPUserPlugin' => $baseDir . '/../lib/ILDAPUserPlugin.php',
     'OCA\\User_LDAP\\ILDAPWrapper' => $baseDir . '/../lib/ILDAPWrapper.php',
     'OCA\\User_LDAP\\IUserLDAP' => $baseDir . '/../lib/IUserLDAP.php',
     'OCA\\User_LDAP\\Jobs\\CleanUp' => $baseDir . '/../lib/Jobs/CleanUp.php',
@@ -47,6 +51,7 @@ return array(
     'OCA\\User_LDAP\\Proxy' => $baseDir . '/../lib/Proxy.php',
     'OCA\\User_LDAP\\Settings\\Admin' => $baseDir . '/../lib/Settings/Admin.php',
     'OCA\\User_LDAP\\Settings\\Section' => $baseDir . '/../lib/Settings/Section.php',
+    'OCA\\User_LDAP\\UserPluginManager' => $baseDir . '/../lib/UserPluginManager.php',
     'OCA\\User_LDAP\\User\\DeletedUsersIndex' => $baseDir . '/../lib/User/DeletedUsersIndex.php',
     'OCA\\User_LDAP\\User\\IUserTools' => $baseDir . '/../lib/User/IUserTools.php',
     'OCA\\User_LDAP\\User\\Manager' => $baseDir . '/../lib/User/Manager.php',
diff --git a/apps/user_ldap/composer/composer/autoload_static.php b/apps/user_ldap/composer/composer/autoload_static.php
index f80900f790..ece7dd69d2 100644
--- a/apps/user_ldap/composer/composer/autoload_static.php
+++ b/apps/user_ldap/composer/composer/autoload_static.php
@@ -39,9 +39,13 @@ class ComposerStaticInitUser_LDAP
         'OCA\\User_LDAP\\Exceptions\\ConstraintViolationException' => __DIR__ . '/..' . '/../lib/Exceptions/ConstraintViolationException.php',
         'OCA\\User_LDAP\\Exceptions\\NotOnLDAP' => __DIR__ . '/..' . '/../lib/Exceptions/NotOnLDAP.php',
         'OCA\\User_LDAP\\FilesystemHelper' => __DIR__ . '/..' . '/../lib/FilesystemHelper.php',
+        'OCA\\User_LDAP\\GroupPluginManager' => __DIR__ . '/..' . '/../lib/GroupPluginManager.php',
         'OCA\\User_LDAP\\Group_LDAP' => __DIR__ . '/..' . '/../lib/Group_LDAP.php',
         'OCA\\User_LDAP\\Group_Proxy' => __DIR__ . '/..' . '/../lib/Group_Proxy.php',
         'OCA\\User_LDAP\\Helper' => __DIR__ . '/..' . '/../lib/Helper.php',
+        'OCA\\User_LDAP\\IGroupLDAP' => __DIR__ . '/..' . '/../lib/IGroupLDAP.php',
+        'OCA\\User_LDAP\\ILDAPGroupPlugin' => __DIR__ . '/..' . '/../lib/ILDAPGroupPlugin.php',
+        'OCA\\User_LDAP\\ILDAPUserPlugin' => __DIR__ . '/..' . '/../lib/ILDAPUserPlugin.php',
         'OCA\\User_LDAP\\ILDAPWrapper' => __DIR__ . '/..' . '/../lib/ILDAPWrapper.php',
         'OCA\\User_LDAP\\IUserLDAP' => __DIR__ . '/..' . '/../lib/IUserLDAP.php',
         'OCA\\User_LDAP\\Jobs\\CleanUp' => __DIR__ . '/..' . '/../lib/Jobs/CleanUp.php',
@@ -62,6 +66,7 @@ class ComposerStaticInitUser_LDAP
         'OCA\\User_LDAP\\Proxy' => __DIR__ . '/..' . '/../lib/Proxy.php',
         'OCA\\User_LDAP\\Settings\\Admin' => __DIR__ . '/..' . '/../lib/Settings/Admin.php',
         'OCA\\User_LDAP\\Settings\\Section' => __DIR__ . '/..' . '/../lib/Settings/Section.php',
+        'OCA\\User_LDAP\\UserPluginManager' => __DIR__ . '/..' . '/../lib/UserPluginManager.php',
         'OCA\\User_LDAP\\User\\DeletedUsersIndex' => __DIR__ . '/..' . '/../lib/User/DeletedUsersIndex.php',
         'OCA\\User_LDAP\\User\\IUserTools' => __DIR__ . '/..' . '/../lib/User/IUserTools.php',
         'OCA\\User_LDAP\\User\\Manager' => __DIR__ . '/..' . '/../lib/User/Manager.php',
```

blizzz commented:
^ commit those please since I cannot push to your repo

viniciuscb commented:
@blizzz: response from git apply: file(s) not found

```
# git apply changes.diff
error: apps/user_ldap/composer/composer/autoload_classmap.php: Arquivo ou diretório não encontrado
error: apps/user_ldap/composer/composer/autoload_static.php: Arquivo ou diretório não encontrado
```

blizzz commented:
@viniciuscb i think it takes a rebase to current master. autoloaders for app have been introduced just recently

viniciuscb commented:
ok! lets see if it work

blizzz commented:
Appears good :) 

I was just informed that freeze was delayed a little bit to next week, having this done earlier is better, not much time buffer anyhow. We need a second reviewer, this should be doable ;)

viniciuscb commented:
A working implementation of User and Group LDAP Plugin. (This works for a very specific LDAP configuration - no guarantees it will work for others).

https://gitlab.com/eita/rios/user_ldap_extended

MorrisJobke reviewed (approved):
Code looks good
